invokeai/app/services/shared/invocation_context.py

```diff
@@ -0,0 +1,625 @@
+
+
+class ModelsInterface:
+    def __init__(self, services: InvocationServices, context_data: InvocationContextData) -> None:
+        def exists(model_name: str, base_model: BaseModelType, model_type: ModelType) -> bool:
+            """
+            Checks if a model exists.
+
+            :param model_name: The name of the model to check.
+            :param base_model: The base model of the model to check.
+            :param model_type: The type of the model to check.
+            """
+            return services.model_manager.model_exists(model_name, base_model, model_type)
+
+        def load(
+            model_name: str, base_model: BaseModelType, model_type: ModelType, submodel: Optional[SubModelType] = None
+        ) -> ModelInfo:
+            """
+            Loads a model, returning its `ModelInfo` object.
+
+            :param model_name: The name of the model to get.
+            :param base_model: The base model of the model to get.
+            :param model_type: The type of the model to get.
+            :param submodel: The submodel of the model to get.
+            """
+
+            # During this call, the model manager emits events with model loading status. The model
+            # manager itself has access to the events services, but does not have access to the
+            # required metadata for the events.
+            #
+            # For example, it needs access to the node's ID so that the events can be associated
+            # with the execution of a specific node.
+            #
+            # While this is available within the node, it's tedious to need to pass it in on every
+            # call. We can avoid that by wrapping the method here.
+
+            return services.model_manager.get_model(
+                model_name, base_model, model_type, submodel, context_data=context_data
+            )
+
+        def get_info(model_name: str, base_model: BaseModelType, model_type: ModelType) -> dict:
+            """
+            Gets a model's info, an dict-like object.
+
+            :param model_name: The name of the model to get.
+            :param base_model: The base model of the model to get.
+            :param model_type: The type of the model to get.
+            """
+            return services.model_manager.model_info(model_name, base_model, model_type)
+
+        self.exists = exists
+        self.load = load
+        self.get_info = get_info
+
+
+class ConfigInterface:
+    def __init__(self, services: InvocationServices) -> None:
+        def get() -> InvokeAIAppConfig:
+            """
+            Gets the app's config. The config is read-only; attempts to mutate it will raise an error.
+            """
+
+            # The config can be changed at runtime.
+            #
+            # We don't want nodes doing this, so we make a frozen copy.
+
+            config = services.configuration.get_config()
+            # TODO(psyche): If config cannot be changed at runtime, should we cache this?
```

Comment:
Cache between nodes of a single graph, or between graphs?

Reply:
I think it makes sense to cache for the life of a graph

Reply:
Because the processor operates on individual nodes, and the context is created anew for each node, I don't think we can easily cache it for the life of a graph. We could either cache it for the lifetime of the app, or an individual node.

Reply:
I've set it up to cache the config. The config is read just-in-time, so we aren't unnecessarily reading the config for every node when that's not needed.

Reply:
Turns out the config class already caches the config.